File: Dev/unitTest_Engine_cs/Graphics/3D/Effect3D.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace unitTest_Engine_cs.Graphics._3D
{
	class Effect3D : EngineTest
	{
		public Effect3D()
			: base(20)
		{
		}

		protected override void OnStart()
		{
			var scene = new asd.Scene();
			var layer = new asd.Layer3D();
			var effectObj = new asd.EffectObject3D();
			var lightObj = new asd.DirectionalLightObject3D();
			var cameraObj = new asd.CameraObject3D();
			var effect = asd.Engine.Graphics.CreateEffect("Data/Effect/Laser01.efk");
			scene.AddLayer(layer);
			layer.AddObject(effectObj);
			layer.AddObject(lightObj);
			layer.AddObject(cameraObj);
			asd.Engine.ChangeScene(scene);

			cameraObj.Position = new asd.Vector3DF(40, 20, 40);
			cameraObj.Focus = new asd.Vector3DF(0, 0, 0);
			cameraObj.FieldOfView = 50.0f;
			cameraObj.ZNear = 1.0f;
			cameraObj.ZFar = 200.0f;
			cameraObj.WindowSize = new asd.Vector2DI(800, 600);

			effectObj.Rotation = new asd.Vector3DF(0.0f, 20.0f, 0.0f);

			lightObj.Rotation = new asd.Vector3DF(30, 160, 0);

			effectObj.SetEffect(effect);
			effectObj.Play();
		}
	}
}

File: Dev/unitTest_Engine_cpp_gtest/Graphics/3D/Effect3D.cpp

#include "../../EngineTest.h"

class Graphics_Effect3D : public EngineTest
{
public:

	Graphics_Effect3D(bool isOpenGLMode) :
		EngineTest(ace::ToAString("Effect3D"), isOpenGLMode, 20)
	{}

protected:
	void OnStart() override
	{
		auto scene = std::make_shared<ace::Scene>();
		auto layer = std::make_shared<ace::Layer3D>();
		auto effectObj = std::make_shared<ace::EffectObject3D>();
		auto lightObj = std::make_shared<ace::DirectionalLightObject3D>();
		auto cameraObj = std::make_shared<ace::CameraObject3D>();
		auto effect = ace::Engine::GetGraphics()->CreateEffect(ace::ToAString("Data/Effect/Laser01.efk").c_str());

		scene->AddLayer(layer);
		layer->AddObject(effectObj);
		layer->AddObject(lightObj);
		layer->AddObject(cameraObj);
		ace::Engine::ChangeScene(scene);

		cameraObj->SetPosition(ace::Vector3DF(40, 20, 40));
		cameraObj->SetFocus(ace::Vector3DF(0, 0, 0));
		cameraObj->SetFieldOfView(50.0f);
		cameraObj->SetZNear(1.0f);
		cameraObj->SetZFar(200.0f);
		cameraObj->SetWindowSize(ace::Vector2DI(800, 600));

		effectObj->SetRotation(ace::Vector3DF(00.0f, 20.0f, 0.0f));

		lightObj->SetRotation(ace::Vector3DF(30, 160, 0));

		effectObj->SetEffect(effect);
		effectObj->Play();
	}
};


TEST(Graphics, Effect3D_GL)
{
	Graphics_Effect3D(true).Run();
	AssertMemoryDoesntLeak();
}

#if _WIN32
TEST(Graphics, Effect3D_DX)
{
	Graphics_Effect3D(false).Run();
	AssertMemoryDoesntLeak();
}
#endif

void Graphics_Effect3D_(bool isGL)
{
	Graphics_Effect3D(isGL).Run();
	AssertMemoryDoesntLeak();
}
